Update expose.md (#152)
percona/k8spsmdb-docs: docs/expose.md

# Exposing the cluster

The Operator provides entry points for accessing the database by client applications in several scenarios. In either way the cluster is exposed with regular Kubernetes [Service objects  :octicons-link-external-16:](https://kubernetes.io/docs/concepts/services-networking/service/), configured by the Operator.

This document describes the usage of [Custom Resource manifest options](operator.md) to expose clusters deployed with the Operator. 

## Using single entry point in a sharded cluster

> If Percona Server for MongoDB [Sharding mode](sharding.md) is turned **on** (default behavior), then database cluster runs special

Reviewer: ```suggestion
If Percona Server for MongoDB [sharding mode](sharding.md) is turned **on** (default behavior), then database cluster runs special
```

`mongos` Pods - query routers, which acts as an entry point for client applications:

![image](assets/images/mongos_espose.png)

> By default, a ClusterIP service is created (this is controlled by [sharding.mongos.expose.exposeType](operator.md#sharding-mongos-expose-exposetype)). The service works in a round-robin fashion between all the `mongos` Pods.

Reviewer: ```suggestion
By default, a ClusterIP Service is created (this is controlled by [sharding.mongos.expose.exposeType](operator.md#sharding-mongos-expose-exposetype)). The Service works in a round-robin fashion between all the `mongos` Pods.
```

> By default, a ClusterIP service is created (this is controlled by [sharding.mongos.expose.exposeType](https://docs.percona.com/percona-operator-for-mongodb/operator.html#shardingmongosexposeexposetype)). The service works in a round-robin fashion between all the `mongos` Pods.

Reviewer: ```suggestion
By default, a ClusterIP service is created (this is controlled by [sharding.mongos.expose.exposeType](operator.md#operator.html#sharding-mongos-expose-exposetype)). The service works in a round-robin fashion between all the `mongos` Pods.
```

The URI looks like this (taking into account the need for a proper password obtained from the Secret, and a proper namespace name instead of the `<namespace name>` placeholder):

``` {.bash data-prompt="$" }
$ mongo "mongodb://userAdmin:userAdminPassword@my-cluster-name-mongos.<namespace name>.svc.cluster.local/admin?ssl=false"
```

You can get the actual Service endpoints by running the following command:

``` {.bash data-prompt="$" }
$ kubectl get psmdb
```

??? example "Expected output"
```
NAME              ENDPOINT                                             STATUS   AGE

> my-cluster-name   my-cluster-name-mongos.test.svc.cluster.local        ready    85m

Reviewer: ```suggestion
my-cluster-name   my-cluster-name-mongos.default.svc.cluster.local        ready    85m
```

> my-cluster-name   my-cluster-name-mongos.default.svc.cluster.local        ready    85m

Reviewer: ```suggestion
my-cluster-name   my-cluster-name-mongos.default.svc.cluster.local     ready    85m
```

```

!!! warning

>     A ClusterIP Service endpoint is only reachable inside Kubernetes. If you need to connect from the outside, you need to expose the mongos Pods by using the NodePort or Load Balancer service types.

Reviewer: ```suggestion
    A ClusterIP Service endpoint is only reachable inside Kubernetes. If you need to connect from the outside, you need to expose the mongos Pods by using the NodePort or Load Balancer Service types.
```

>     See the "Connecting from outside Kubernetes" section below for details.

Reviewer: ```suggestion
    See the [Connecting from outside Kubernetes](expose.md#connecting-from-outside-kubernetes) section below for details.
```

therefore binding communication between Pods to specific IP addresses would
cause problems as things change over time as a result of the cluster scaling,
maintenance, etc. Due to this changing environment, you should connect to
Percona Server for MongoDB by using Kubernetes internal DNS names in the URI.

> By default, a ClusterIP Service is created (this is controlled by [replsets.expose.exposeType](https://docs.percona.com/percona-operator-for-mongodb/operator.html#replsetsexposeexposetype)). The service works in a round-robin fashion between all the mongod Pods of the replica set.

Reviewer: ```suggestion
By default, a ClusterIP type Service is created (this is controlled by [replsets.expose.exposeType](https://docs.percona.com/percona-operator-for-mongodb/operator.html#replsetsexposeexposetype)). The service works in a round-robin fashion between all the mongod Pods of the replica set.
```

In this case, the URI looks like this (taking into account the need for a proper password obtained from the Secret, and a proper namespace name instead of the `<namespace name>` placeholder):

``` {.bash data-prompt="$" }
$ mongodb://databaseAdmin:databaseAdminPassword@my-cluster-name-rs0.<namespace name>.svc.cluster.local/admin?replicaSet=rs0&ssl=false"
```

You can get the actual Service endpoints by running the following command:

``` {.bash data-prompt="$" }
$ kubectl get psmdb
```

??? example "Expected output"
```
NAME              ENDPOINT                                             STATUS   AGE

> minimal-cluster   my-cluster-name-rs0.test.svc.cluster.local           ready    2m19s

Reviewer: ```suggestion
my-cluster-name   my-cluster-name-rs0.default.svc.cluster.local           ready    2m19s
```

> my-cluster-name   my-cluster-name-rs0.default.svc.cluster.local           ready    2m19s

Reviewer: ```suggestion
my-cluster-name   my-cluster-name-rs0.default.svc.cluster.local        ready    2m19s
```

```
!!! warning

>     A ClusterIP service endpoint is only reachable inside Kubernetes. If you need to connect from the outside, you need to expose the mongod pods by using the NodePort or Load Balancer service types.

Reviewer: ```suggestion
    A ClusterIP service endpoint is only reachable inside Kubernetes. If you need to connect from the outside, you need to expose the mongod Pods by using the NodePort or Load Balancer Service types.
```

therefore binding communication between Pods to specific IP addresses would
cause problems as things change over time as a result of the cluster scaling,
maintenance, etc. Due to this changing environment, you should connect to
Percona Server for MongoDB by using Kubernetes internal DNS names in the URI.

> By default, a ClusterIP type Service is created (this is controlled by [replsets.expose.exposeType](https://docs.percona.com/percona-operator-for-mongodb/operator.html#replsetsexposeexposetype)). The service works in a round-robin fashion between all the mongod Pods of the replica set.

Reviewer: ```suggestion
By default, a ClusterIP type Service is created (this is controlled by [replsets.expose.exposeType](operator.md#replsets-expose-exposetype)). The Service works in a round-robin fashion between all the mongod Pods of the replica set.
```

In this case, the URI looks like this (taking into account the need for a proper password obtained from the Secret, and a proper namespace name instead of the `<namespace name>` placeholder):

``` {.bash data-prompt="$" }
$ mongodb://databaseAdmin:databaseAdminPassword@my-cluster-name-rs0.<namespace name>.svc.cluster.local/admin?replicaSet=rs0&ssl=false"
```

You can get the actual Service endpoints by running the following command:

``` {.bash data-prompt="$" }
$ kubectl get psmdb
```

??? example "Expected output"
```
NAME              ENDPOINT                                             STATUS   AGE
my-cluster-name   my-cluster-name-rs0.default.svc.cluster.local        ready    2m19s
```
!!! warning

>     A ClusterIP service endpoint is only reachable inside Kubernetes. If you need to connect from the outside, you need to expose the mongod Pods by using the NodePort or Load Balancer Service types.

Reviewer: ```suggestion
    A ClusterIP Service endpoint is only reachable inside Kubernetes. If you need to connect from the outside, you need to expose the mongod Pods by using the NodePort or Load Balancer Service types.
```

>     See the "Connecting from outside Kubernetes" section below for details.

Reviewer: ```suggestion
    See the [Connecting from outside Kubernetes](expose.md#connecting-from-outside-kubernetes) section below for details.
```

    
## Connecting from outside Kubernetes

If connecting to a cluster from outside Kubernetes, you cannot reach the Pods using the Kubernetes internal DNS
names. To make the Pods accessible, Percona Operator for MongoDB
can create [Kubernetes Services  :octicons-link-external-16:](https://kubernetes.io/docs/concepts/services-networking/service/).

> * set `expose.enabled` option to `true` to allow exposing the Pods via services,

Reviewer: ```suggestion
* set `expose.enabled` option to `true` to allow exposing the Pods via Services,
```

* set `expose.exposeType` option specifying the type of Service to be used:
    * `ClusterIP` - expose the Pod with an internal static
        IP address. This variant makes the Service reachable only from
        within the Kubernetes cluster.

>     * `NodePort` - expose the Pod on each Kubernetes node’s

Reviewer: ```suggestion
    * `NodePort` - expose the Pod on each Kubernetes Node’s
```

        IP address at a static port. A ClusterIP Service, to which the Node
        port will be routed, is automatically created in this variant. As

>         an advantage, the service will be reachable from outside the

Reviewer: ```suggestion
        an advantage, the Service will be reachable from outside the
```

>         cluster by node address and port number, however the address will be

Reviewer: ```suggestion
        cluster by Node address and port number, however the address will be
```

>         bound to a specific Kubernetes node.

Reviewer: ```suggestion
        bound to a specific Kubernetes Node.
```

    * `LoadBalancer` - expose the Pod externally using a
        cloud provider’s load balancer. Both [ClusterIP and NodePort
        Services are automatically created :octicons-link-external-16:](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer) in this variant

If the NodePort type is used, the URI looks like this:

```mongodb://databaseAdmin:databaseAdminPassword@<node1>:<port1>,<node2>:<port2>,<node3>:<port3>/admin?replicaSet=rs0&ssl=false```

> All node adresses should be *directly* reachable by the application.

Reviewer: ```suggestion
All Node addresses should be *directly* reachable by the application.
```

## Service per Pod

> To make the Pods accessible, Percona Operator for MongoDB can assign a [Kubernetes Service  :octicons-link-external-16:](https://kubernetes.io/docs/concepts/services-networking/service/) to each Pod.

Reviewer: ```suggestion
To make all database Pods accessible, Percona Operator for MongoDB can assign a [Kubernetes Service  :octicons-link-external-16:](https://kubernetes.io/docs/concepts/services-networking/service/) to each Pod.
```

> The Service per Pod option is available to allow the application to take care of Cursor tracking instead of relying on a single Service. This solves the

Reviewer: ```suggestion
Particularly, the Service per Pod option allows the application to take care of Cursor tracking instead of relying on a single Service. This solves the
```

problem of CursorNotFound errors when the Service transparently cycles between the mongos instances while client is still iterating the cursor
on some large collection.

> This feature can be configured in the [sharding.mongos.expose.servicePerPod](operator.md#sharding-mongos-expose-serviceperpod) (for mongos Pods) sections of the

Reviewer: ```suggestion
This feature can be turned on by setting the [sharding.mongos.expose.servicePerPod](operator.md#sharding-mongos-expose-serviceperpod) Custom Resource option to `true`.
```

> This feature can be configured in the [sharding.mongos.expose.servicePerPod](https://docs.percona.com/percona-operator-for-mongodb/operator.html#shardingmongosexposeserviceperpod) (for mongos Pods) sections of the

Reviewer: ```suggestion
This feature can be configured in the [sharding.mongos.expose.servicePerPod](operator.md#sharding-mongos-expose-serviceperpod) (for mongos Pods) sections of the
```

[deploy/cr.yaml  :octicons-link-external-16:](https://github.com/percona/percona-server-mongodb-operator/blob/main/deploy/cr.yaml) file.

> If this feature is enabled with the exposeType: NodePort, the created services look like this:

Reviewer: ```suggestion
If this feature is enabled with the `exposeType: NodePort`, the created Services look like this:
```
